Sources/SwiftParser/Diagnostics/ParserDiagnosticMessages.swift

```diff
@@ -114,11 +128,42 @@ public struct ExtaneousCodeAtTopLevel: ParserError {
   }
 }
 
+public struct MissingNodeError: ParserError {
+  public let missingNode: Syntax
+
+  public var message: String {
+    var message: String
+    var hasNamedParent = false
+    if let parent = missingNode.parent,
+        let childName = parent.childNameForDiagnostics(missingNode.index) {
+      message = "Expected \(childName)"
+      if let parent = missingNode.parent,
+          let parentTypeName = parent.nodeTypeNameForDiagnostics(inherit: false) {
+        message += " of \(parentTypeName)"
+        hasNamedParent = true
+      }
+    } else {
+      message = "Expected \(missingNode.nodeTypeNameForDiagnostics() ?? "syntax")"
+      if let lastChild = missingNode.lastToken(viewMode: .fixedUp), lastChild.presence == .present {
+        message += " after '\(lastChild.text)'"
+      } else if let previousToken = missingNode.previousToken(viewMode: .fixedUp), previousToken.presence == .present {
+        message += " after '\(previousToken.text)'"
```

Comment:
> If `previousToken.parent != missingNode.parent`, could we use the parent of the previous token rather than the token itself? That would give both the attribute tests `Expected declaration after attribute` rather than empty/')', but still keep `after <token>` for cases where it makes sense (eg. `a ? b :`).

`after attribute` should now be added based.

> Actually speaking of the ternary case, shouldn't it have `in ternary expression` at the end of it?

It doesn’t have `in ternary expression` at the end of it because the expression is not missing inside the ternary expression but in the sequence expression. For `a ? b :` the syntax tree looks like
- IdentifierExpr: `a`
- UnresolvedTernaryExpr: `? b :`
- MissingExpr